Trust committed ops

jorangreef opened the pull request:
Fixes: #100

jorangreef committed:
VSR: Squash of trust_committed_ops

VSR: Send commit_min in do_view_change message

Instead of sending commit_max, which is not strictly required.

Sending commit_min enables the new primary to determine whether
it must trust and therefore repair using another replica's op.

For example, in situations where the hash chain has a gap that
would otherwise prevent the repair, leading to a log fork.

VSR: Document why prepare_timestamp only advances

VSR: Add TODO for discard_uncommitted_headers()

Refs: https://github.com/coilhq/tigerbeetle/issues/100

VSR: Improve comments

VSR: Fix repair_header() to not introduce stale headers

- Do not introduce any header if not fully connected to the hash chain head.
- Even if the header exists as dirty (so as not to introduce I/O concurrency).
- Fix incorrect assertion that an existing header has the same op number.
- Improve debug logs to include checksum.
- Use simpler journal.has() over header_with_op_and_checksum().
- Rename journal.header_for_entry() to header_for_prepare().
- Remove out-of-dae comment regarding overlap.
- Fix line lengths.

VSR: Remove repair_header_would_break_hash_chain_with_next_entry()

VSR: Fix on_do_view_change() and discard_uncommitted_headers()

- Fix on_do_view_change() to replace canonical/committed headers.
- Fix discard_uncommitted_headers() to discard any gaps in the pipeline.
- Refactor on_do_view_change() to be under 70 lines.
- Refactor on_start_view(), on_recovery_response() to use replace_headers().

TODO Must on_do_view_change() first erase the portion of the log covered
by all the do_view_change messages, so that discard_uncommitted_headers()
will then see any local stale headers also as gaps?

VSR: Apply code review suggestions

- Do not replace a header if committed locally or from a previous log cycle.
- Use consistent statement order when rewinding replica.op.
- Use const instead of var.
- Improve documentation.
- Add assertions.

VSR: Add function documentation for replace_header()

VSR: Discard uncanonical headers before installing quorum headers

VSR: Add TODO to optimize discard_uncommitted_headers()

VSR: Strength op_highest() to verify that headers are continuous

VSR: Verify a replica's do_view_change message when installing quorum

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -2545,7 +2573,8 @@ pub fn Replica(
                 .op = self.op,
-                .commit = self.commit_max,
+                // See the comment in `on_do_view_change()` for why `commit_min` is crucial:
+                .commit = if (command == .do_view_change) self.commit_min else self.commit_max,
```
`Header.commit`'s doc comment in `vsr.zig` should document this variation, e.g.:

```zig
    /// The commit number of the latest committed prepare. Committed ops are immutable.
    ///
    /// * A `do_view_change` sets this to `commit_min`.
    /// * A `start_view` sets this to `commit_max`.
    commit: u64 = 0,
```

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4297,7 +4326,14 @@ pub fn Replica(
+            // the primary also discards uncommitted operations.
+            // It is also safe not to use `commit_max` here because the new primary will assume that
+            // operations after the highest `commit_min` may yet have been committed before the old
+            // primary crashed. The new primary will use the NACK protocol to be sure of a discard.
```
Maybe `s/repaired in its own log/replaced in its own log`?
The headers we are talking about here aren't installed by `repair_header`, so "repair" might be a little misleading.

jorangreef replied:
I was thinking the same, wasn't quite happy with overloading the repair wording.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -1070,10 +1070,24 @@ pub fn Replica(
-                        _ = self.repair_header(h);
+                        if (h.op <= m.header.commit) {
+                            if (self.journal.header_with_op_and_checksum(h.op, h.checksum)) |_| {
+                                log.debug("{}: on_do_view_change: op={} committed on replica={}", .{
```
`op={} committed on replica={}` could describe the `else` branch also.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -1085,14 +1099,23 @@ pub fn Replica(
+            // TODO At present, discard_uncommitted_headers() does not consider interactions with
+            // the hash chain, i.e. that could lead to a hash chain swing, when it removes
+            // uncommitted ops. This is a correctness bug, where old ops may not be replaced in the
+            // view change. Rather, discard_uncommitted_headers() must fix up self.op after discard.
```
> that could lead to a hash chain swing

I'm not sure I understand what a "hash chain swing" is.

I understand that the new `self.journal.set_header_as_dirty` call above can replace an op, potentially breaking the hash chain between `commit_min` and `commit_min + 1` (for every `do_view_change`'s `commit_min`). So we may need to discard some ops that follow a `commit_min`.

But the comment implies the "swing" hasn't happened yet?

jorangreef replied:
A “hash chain swing” or “quorum swing” would be where a prepare is replaced with another, undermining the guarantee to the previous primary, meaning that the commit quorum “swings” to a new decision by another primary, forking the log if the prepare was in fact committed.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -1024,7 +1024,7 @@ pub fn Replica(
 
             var v: ?u32 = null;
-            var k: ?u64 = null;
+            var k: ?u64 = null; // This represents `commit_min` according to each replica.
```
Should it be

```
// The highest `commit_min` from any replica.
```

jorangreef replied:
Much better!

jorangreef commented:
Fixes: #103

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -2585,105 +2520,52 @@ pub fn Replica(
                 log.debug("{}: discard_uncommitted_headers: op={} gap", .{ self.replica, op });
 
-                var nacks: usize = 0;
-                for (self.do_view_change_from_all_replicas) |received, replica| {
```
It's great that we can remove the nack protocol; this really simplifies the view change!

jorangreef replied:
Yes, much better! Just to clarify that we only remove it here. We still need it later, for discarding ops where there are no gaps.

However, using it here was never necessary for correctness, since the view change already guarantees the quorum intersection property, i.e. if any header might have committed, then it is guaranteed to survive into the view change. Therefore, if a header does not survive into a view change, if there is a gap across all replicas, then that op (and subsequent ops) could not possibly have committed. So the mere existence of a gap is enough to discard here.

`discard_uncommitted_headers()` is trying to discard headers for which there are gaps, where we are certain the op could not have been committed.

The NACK protocol has a slightly different purpose: discern for sure whether an op (for which there is no gap, where the header exists) is definitely uncommitted, and can therefore be discarded safely if it cannot be repaired.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4296,8 +4170,15 @@ pub fn Replica(
+            // the primary also discards uncommitted operations.
+            // It is also safe not to use `commit_max` here because the new primary will assume that
+            // operations after the highest `commit_min` may yet have been committed before the old
+            // primary crashed. The new primary will use the NACK protocol to be sure of a discard.
```
> The new primary will use the NACK protocol to be sure of a discard.

Not anymore, right?

jorangreef replied:
This is still true, since we have two ways of discarding: for a clear gap, or otherwise via the NACK protocol. We just no longer use the NACK protocol where there is already a clear gap.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4086,6 +3951,15 @@ pub fn Replica(
         }
 
+        fn replace_header(self: *Self, header: *Header) void {
+            assert(header.command == .prepare);
```
Could we assert that `header.op <= self.op` to verify that the caller has already updated the replica's `op`?

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4086,6 +3951,15 @@ pub fn Replica(
+            // Do not set an op as dirty if we already have it exactly because:
+            // 1. this would trigger a repair and delay the view change, or worse,
+            // 2. prevent repairs to another replica when we have the op.
+            if (!self.journal.has(header)) self.journal.set_header_as_dirty(header);
```
Inside this condition, could we assert that if we are replacing another version of the same op, then `op > self.commit_min`?
Or maybe we could assert that `replace_header` is never called when `op ≤ commit_min`.

jorangreef replied:
I tried the first assert, and then realized that it might also be possible that our journal may simply be ahead and may not have the op any more? So an early return, following `repair_header()`, is what we have now in `replace_header()` if the header being replaced is already locally committed.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4479,112 +4360,195 @@ pub fn Replica(
-                log.debug("{}: {s}: latest op exists exactly", .{ self.replica, method });
-            } else {
-                self.journal.set_header_as_dirty(latest);
+            self.journal.remove_entries_from(self.op + 1);
```
Could this move up to immediately follow `self.op = op`? It's helpful when those are paired close together.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4479,112 +4360,195 @@ pub fn Replica(
+            // First, set all the canonical headers from the replica(s) with highest `view_normal`:
+            for (self.do_view_change_from_all_replicas) |received| {
+                if (received) |message| {
+                    var view_normal = @intCast(u32, message.header.timestamp);
```
```suggestion
                    const view_normal = @intCast(u32, message.header.timestamp);
```

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4479,112 +4360,195 @@ pub fn Replica(
+            // Now that the canonical headers are all in place, repair any other headers:
+            for (self.do_view_change_from_all_replicas) |received| {
+                if (received) |message| {
+                    var view_normal = @intCast(u32, message.header.timestamp);
```
```suggestion
                    const view_normal = @intCast(u32, message.header.timestamp);
```

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4479,112 +4360,195 @@ pub fn Replica(
+            for (self.do_view_change_from_all_replicas) |received| {
+                if (received) |message| {
+                    var view_normal = @intCast(u32, message.header.timestamp);
+                    assert(view_normal < message.header.view);
```
Can you explain this assertion to me? (Why can't they be equal?)

jorangreef replied:
The view during which the replica's status was last normal, has to be before the view of the current view change, since we are advancing to a newer view. Is that a clear explanation?

I've added a comment to this effect.

sentientwaffle replied:
Gotcha, this message is always generated during a view change, which means it's view_normal is in the past. I was mixed up about the context.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4296,8 +4170,15 @@ pub fn Replica(
+            // It is also safe not to use `commit_max` here because the new primary will assume that
+            // operations after the highest `commit_min` may yet have been committed before the old
+            // primary crashed. The new primary will use the NACK protocol to be sure of a discard.
+            assert(message.header.commit == self.commit_min);
```
Copy this assertion to `send_message_to_replica`'s `.do_view_change` section.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4296,8 +4170,15 @@ pub fn Replica(
-            assert(message.header.op == self.message_body_as_headers(message)[0].op);
-            assert(message.header.commit == self.commit_max);
+            assert(message.header.op == message_body_as_headers(message)[0].op);
+            // Each replica must advertise its own commit number, so that the primary can know which
```
```suggestion
            // Each replica must advertise its own commit number, so that the new primary can know which
```

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -2520,6 +2520,49 @@ pub fn Replica(
+            assert(remove_op_min > self.commit_min);
+            assert(remove_op_min + config.pipeline_max - 1 >= self.op);
+
+            // If any ops are canonical, then the replica's do_view_change message will repair them.
```
Could the replica update `self.op` here, just for consistency? (I know that it is going to be immediately rewritten by `set_op_and_commit_max`).

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -2520,6 +2520,49 @@ pub fn Replica(
+            assert(remove_op_count <= config.pipeline_max);
+            if (remove_op_count == 0) return;
+
+            const remove_op_min = self.op -| (remove_op_count - 1); // Inclusive.
```
Is the saturating subtraction necessary?

- When `remove_op_count = self.op - self.commit_min`, `self.op - (remove_op_count - 1) = (self.op - remove_op_count) + 1 = commit_min + 1`.
- When `remove_op_count = config.pipeline_max`, `config.pipeline_max < self.op - self.commit_min < self.op`.

jorangreef replied:
Thanks, the saturating arithmetic is not necessary. This would also make a great comment to show the workings.

There's a small tweak I think.

- When `remove_op_count = config.pipeline_max`, `config.pipeline_max < self.op - self.commit_min` holds.
- However, from there `self.op - self.commit_min < self.op` does not always follow. For example, when `self.op=1` and `commit_min=0` then `self.op - self.commit_min == self.op`.

jorangreef replied:
This is what I have:

```zig
const remove_op_count = std.math.min(
    // Do not reset any ops that we have already committed.
    self.op - self.commit_min,
    // The number of uncommitted ops cannot be more than the length of the pipeline.
    // Do not reset any ops that we did not include in our do_view_change message.
    config.pipeline_max,
);

assert(remove_op_count <= config.pipeline_max);
if (remove_op_count == 0) return;

// Saturating subtraction is not necessary below, because we take the min() above:
// * When `remove_op_count` = `self.op - self.commit_min`,
//      `self.op - self.commit_min <= self.op`.
// * When `remove_op_count = config.pipeline_max`,
//      `config.pipeline_max < self.op - self.commit_min <= self.op`.
const remove_op_min = self.op - remove_op_count + 1; // Add 1 as `op_min` is inclusive.
```

However, I think this is harder to get wrong, as it also binds intention with respect to the replica's DVC:

```zig
const message = self.do_view_change_from_all_replicas[self.replica].?;
assert(message.header.view == self.view);
assert(message.header.op == self.op);

var op_min: ?u64 = null;
var op_count: usize = config.pipeline_max;

for (message_body_as_headers(message)) |header| {
    assert(op_min == null or op_min.? > header.op);

    if (header.op <= self.commit_min) continue;
    if (op_count == 0) continue; // Assert all subsequent header.op's against op_min.

    op_min = header.op;
    op_count -= 1;
}
```

Now we have a stronger guarantee that any header we discard here as potentially uncanonical, will be able to be repaired later from the replica's DVC, if it turns out to be canonical.

sentientwaffle replied:
Can we add
```
assert(message.header.commit == self.commit_min);
```
to the `message`'s assertions?

---

`op_min` is a tricky name, since it is about to become the new `replica.op` (which is the replica's _maximum_ op). What about `op_uncanonical_min` or `op_truncate`?

---

Does the `op_count` decrement track correctly when there are gaps in the DVC headers?

sentientwaffle replied:
Also, when the `if (op_count == 0)` branch is taken, doesn't that mean that more than `config.pipeline_max` ops will be removed? i.e. shouldn't it be

```zig
if (op_count == 0) {
    op_min = op_min.? + 1;
    continue
}
```

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -1023,76 +1025,13 @@ pub fn Replica(
 
-            self.discard_uncommitted_headers();
+            self.set_log_from_do_view_change_messages();
             assert(self.op >= self.commit_max);
```
```suggestion
            assert(self.op >= self.commit_max);
            assert(self.state_machine.prepare_timestamp >= self.journal.header_with_op(self.op).?.timestamp);
```

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4479,112 +4413,205 @@ pub fn Replica(
+            });
+        }
+
+        fn set_log_from_do_view_change_messages(self: *Self) void {
```
```suggestion
        fn set_log_from_do_view_change_messages(self: *Self) void {
            assert(self.status == .view_change);
            assert(self.leader_index(self.view) == self.replica);
            assert(self.replica_count > 1);
            assert(self.start_view_change_quorum);
            assert(self.do_view_change_quorum);

```

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4479,112 +4413,205 @@ pub fn Replica(
+        /// Returns the highest `view_normal` from any replica, whose headers will be canonical,
+        /// since the replica has knowledge of previous view changes in which headers were replaced.
+        fn set_op_and_commit_max_from_do_view_change_messages(self: *Self) u64 {
+            assert(self.status == .view_change);
```
```suggestion
            assert(self.status == .view_change);
            assert(self.leader_index(self.view) == self.replica);
```

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4479,112 +4413,205 @@ pub fn Replica(
+            // Any headers removed must be a subset of the replica's do_view_change message, and
+            // must be removed before set_op_and_commit_max() below, since that may change self.op.
+            self.discard_uncanonical_headers(v.?);
+            self.set_op_and_commit_max(n.?, k.?, "on_do_view_change");
```
Since `self.set_op_and_commit_max` here is advancing `self.op`, after this runs `self.journal.header_for_op(self.op)` would be `null` — which temporarily violates a replica invariant.

It might be cleaner if:
- `set_op_and_commit_max` moved into `set_log_from_do_view_change_messages` immediately before or after the first `replace_header` loop
- move `discard_uncanonical_headers`  into `set_log_from_do_view_change_messages` as well, so that all mutation would happen in one place
- Then what's left of `set_op_and_commit_max_from_do_view_change_messages` function simply returns `struct { view_normal, op, commit }` (it would be renamed).

sentientwaffle committed:
WAL: Zero sector padding in read_prepare fast path

In the `read_prepare` fast-path (`prepare.header.size == @sizeOf(Header)`) we don't have to read from disk — the header is already in memory. But we still need to zero out the sector padding so that the `Debug` mode verification succeeds.

Encountered in seed `8320396198111625356`.

sentientwaffle committed:
VSR: Fix trust-committed-ops

- `recovery_response` must bundle at least as many headers as a DVC.
  - (Explained in code: see comment `A recovery response attaches at least as many headers as a DVC message attaches.`).
  - Explain this invariant for the DVC quorum as well (`For each DVC in the quorum`).
- Don't discard uncanonical headers until just before `discard_uncommitted_headers`.
  - Discarding the uncanonical headers means that when the headers are restored from the DVC's headers, some formerly "clean" headers are now "dirty".
  - This harms availability, since we effectively "forget" that we already wrote a prepare.
- Add a `replace_headers` function for loading canonical headers with some extra verification.
- Add additional assertions & guards to `replace_header`.
- Simplify `discard_uncommitted_headers`: truncate when the first gap is found.
- Document why iteration order of DVCs does _not_ impact the final result, because this was unintuitive to me.

jorangreef commented on src/config.zig:
```diff
@@ -95,7 +95,8 @@ pub const connections_max = replicas_max + clients_max;
     // Use a small message size during the simulator for improved performance.
-    .simulation => sector_size,
+    // (*2 so that there is room for pipeline_max headers in the DVC).
+    .simulation => sector_size * 2,
```
Could we derive this exactly, taking the max?

jorangreef replied:
So that we require a min of `pipeline_max` headers, and then round up to the nearest sector? This will leave `pipeline_max` more future proof, especially when we increase `clients_max`.

jorangreef commented on src/vsr.zig:
```diff
@@ -199,6 +199,9 @@ pub const Header = extern struct {
 
     /// The commit number of the latest committed prepare. Committed ops are immutable.
+    ///
+    /// * A `do_view_change` sets this to `commit_min`.
```
```suggestion
    /// * A `do_view_change` sets this to `commit_min`, to reflect the sending replica's progress.
```

jorangreef commented on src/vsr/journal.zig:
```diff
@@ -725,6 +727,7 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
             if (self.header_with_op_and_checksum(op, checksum)) |exact| {
                 if (exact.size == @sizeOf(Header)) {
                     message.header.* = exact.*;
+                    std.mem.set(u8, message.buffer[@sizeOf(Header)..config.sector_size], 0);
```
Can you help me understand why we need to do this?

jorangreef replied:
It looks interesting! 😁

jorangreef replied:
My guess is that it's in case this message is received locally and persisted to disk, so that we don't leak random data into the sector padding? I think usually the MessageBus would otherwise take care of zeroing the padding here?

jorangreef replied:
We could also assert this in debug builds or for `config.verify`, at call sites where we write this message to disk.

sentientwaffle replied:
Yeah, this isn't directly related to trust-committed-ops, it is a just a bug I ran into during testing!

It is documented in the commit https://github.com/tigerbeetledb/tigerbeetle/pull/102/commits/53e7d106e1f60751e54ae8c2685a689931807a53. But I'll add an explanation in the code, too.

> We could also assert this in debug builds or for `config.verify`, at call sites where we write this message to disk.

We do — that's what caught this error :grin:.

jorangreef replied:
Nice! I had a feeling we had those asserts in there somewhere but thought I'd pretend we didn't just in case!

jorangreef replied:
😁

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -63,6 +63,21 @@ const quorum_messages_null = [_]?*Message{null} ** config.replicas_max;
+// that cannot be repaired because they are gaps, and this must be relative to the
+// cluster as a whole (not relative to the difference between our op and commit number)
+// as otherwise we would break correctness.
+const view_change_headers_count = config.pipeline_max;
```
Great to capture it like this.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -877,6 +892,8 @@ pub fn ReplicaType(
             if (message.header.op > self.op + 1) {
                 log.debug("{}: on_prepare: newer op", .{self.replica});
                 self.jump_to_newer_op_in_normal_status(message.header);
+                // "`replica.op` exists" invariant is temporarily broken.
+                assert(self.journal.header_with_op(message.header.op - 1) == null);
```
What if we have a header shining through here from a previous log cycle?

sentientwaffle replied:
We might, but it wouldn't be returned by `header_with_op()` (which only returns an exact `op` match).

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1106,6 +1125,21 @@ pub fn ReplicaType(
         /// ⟨start_view v, l, n, k⟩ messages to the other replicas, where l is the new log, n is the
         /// op number, and k is the commit number.
+        ///
+        /// For each DVC in the quorum:
+        ///
+        /// * The headers must all belong to the same hash chain. (Gaps are allowed).
+        ///   (Reason: the headers bundled with the DVC(s) with the highest view_normal will be
+        ///   loaded into the new leader with `replace_header()`, not `repair_header()`).
+        ///
+        /// Across all DVCs in the quorum:
+        ///
+        /// * The headers of every DVC with the same view_normal must agree. In other words:
+        ///   dvc₁.headers[i].op == dvc₂.headers[j].op implies
+        ///   dvc₁.headers[i].checksum == dvc₂.headers[j].checksum.
+        ///   (Reason: the headers bundled with the DVC(s) with the highest view_normal will be
+        ///   loaded into the new leader with `replace_header()`, not `repair_header()`).
```
These are important invariants to list. I had to do a double-take for the 2nd invariant but it checks out.

sentientwaffle replied:
It was surprising to me too! I "discovered" it because it wasn't being completely enforced — it could be broken by recovery protocol, which didn't include enough headers in the `recovery_response` message.

sentientwaffle committed:
VSR: Document a DVC anti-invariant

This came about after investigating the Journal TODO:

    // TODO Use 2 separate header lists: "staging" and "working".
    // When participating in a view change, each replica should only send the headers from its
    // working set that it knows it prepared.

The split is still necessary due to the second part of the TODO:

    // This also addresses the problem of redundant headers being written prematurely due to
    // batching (after the first log cycle — for the first log cycle we write an invalid message).

(The TODO will be removed in the other branch, where this change is implemented).

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1345,7 +1299,40 @@ pub fn ReplicaType(
+            //    pick 5a — 5a may be committed by replica 0.
+            //    Without replica 0's assistance, replica 1 has no way to pick between 5a/5b.
+            //
+            // Including at least as many headers in the recovery response as the DVC maintains the
+            // invariant: DVCs with the same view_normal must never disagree on the identity of a
+            // message.
+            //
+            // (DVCs can still safely include gaps — but they must be of the form [4a,__,6a],
+            // not [4a,__,6b]).
```
🥳

jorangreef replied:
Awesome worked example.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1345,7 +1299,40 @@ pub fn ReplicaType(
+            //
+            // (DVCs can still safely include gaps — but they must be of the form [4a,__,6a],
+            // not [4a,__,6b]).
+            const count_max = view_change_headers_count;
```
To avoid introducing another local variable, i.e. to rather "shrink the scope", let's just use `view_change_headers_count` inline below when passing to `copy_latest...`, especially since we only use it once.

Then we just add a comma after the last arg to `copy_latest...` to deal with line length.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2911,16 +2901,11 @@ pub fn ReplicaType(
-            // cluster as a whole (not relative to the difference between our op and commit number)
-            // as otherwise we would break correctness.
-            const count_max = config.pipeline_max;
+            const count_max = view_change_headers_count;
```
Here too let's use `view_change_headers_count` inline since we only use it essentially once (`copy_latest...` will also do the `> 0` assert).

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4491,6 +4482,56 @@ pub fn ReplicaType(
+            assert(header.command == .prepare);
+            assert(header.op <= self.op); // Never advance the op.
+            // TODO(Rebase) assert that it is ≤ the checkpoint trigger
+            //assert(header.op <= self.op_checkpoint_trigger());
```
Should this assert be re-enabled now?

sentientwaffle replied:
Yes, thank you!

jorangreef replied:
We should improve our git precommit hook to lint the code for any `// assert` patterns.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4491,6 +4482,56 @@ pub fn ReplicaType(
+                        return;
+                    } else {
+                        // If an op is committed but not checkpointed, we must still have the header.
+                        @panic("missing committed, uncheckpointed header");
```
Nice!

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            //
+            // Don't remove the uncanonical headers yet, though — even though the removed headers are
+            // a subset of the DVC headers, removing and then adding them back would cause clean
+            // headers to become dirty.
```
Good catch!

jorangreef replied:
"yet, though" -> "yet"

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            const op_uncanonical = self.op_uncanonical_min(view_normal_canonical);
+            assert(op_uncanonical <= self.op + 1);
+            assert(op_uncanonical >= self.op + 1 -| config.pipeline_max);
+            assert(op_uncanonical >= self.commit_min);
```
Should it not always be > `self.commit_min`?

jorangreef replied:
i.e. Can `commit_min` itself be uncanonical?

sentientwaffle replied:
Good point; `commit_min` is always canonical. `op_uncanonical_min` already had the correct check, too.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            //
+            // (If replica 0's view_normal was greater than 1/2's, then replica 0 must have all
+            // headers from previous views. Which means 6,7 are from the current view. But since
+            // replica 0 doesn't have 6/7, then replica 1/2 must share the latest view_normal. ∎)
```
```suggestion
            // replica 0 doesn't have 6/7, then replica 1/2 must share the latest view_normal.)
```

A microdot? 😉

jorangreef replied:
Thanks for the awesome worked example!

sentientwaffle replied:
That is a deliberate ["end-of-proof"](https://en.wikipedia.org/wiki/Tombstone_(typography)) symbol, though I can remove it if you prefer.

jorangreef replied:
Let's keep it. And then let it be an Easter egg. i.e. We have so many proofs that we won't always use it.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            assert(self.journal.header_with_op(self.op) != null);
+
+            // The iteration order of DVCs for repair does not impact the final result.
+            // In other words, you can't end up in a situation with a DVC quorums like:
```
```suggestion
            // In other words, you can't end up in a situation with a DVC quorum like:
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            //         3  (4 5 6 7 8)          8 (didn't participate in view change)
+            //         4  (4 5 6 7 8)          8 (didn't participate in view change)
+            //
+            // where the new leader's headers depends on which of replica 1 and 2's DVCs is used
```
```suggestion
            // where the new leader's headers depends on which of replica 1 and 2's DVC is used
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            // 1. is advanced if behind the cluster, but never reset if ahead of the cluster, i.e.
+            // 2. may not always reflect the timestamp of the latest prepared op, and
+            // 3. should be advanced before discarding the timestamps of any uncommitted headers.
+            const head = self.journal.header_with_op(self.op).?;
```
```suggestion
            const head_timestamp = self.journal.header_with_op(self.op).?.timestamp;
```

To shrink the scope further for purpose.

jorangreef replied:
We could also consider adding `timestamp` to `do_view_change_quorum_head` since that is what we are thinking about here... does any quorum have a newer prepare timestamp?

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            /// the replica has knowledge of previous view changes in which headers were replaced.
+            view_normal: u32,
+            /// The highest `commit_min` from any DVC.
+            /// (This is not a `commit_max`.)
```
```suggestion
            /// The highest `commit_min` from any DVC (this is not a `commit_max`).
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+                        assert(view_normal == self.view_normal);
+                        assert(message.header.op == self.op);
+                        // We may have a newer commit than our own DVC due to async commits
+                        // (see below).
```
```suggestion
                        // We may have a newer commit than our DVC due to async commits (see below).
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            // 2. Send `do_view_change` to self.
+            // 3. Finish committing op=N.
+            // 4. Remaining `do_view_change` messages arrive, completing the quorum.
+            // In this scenario, the our own `do_view_change`'s commit is `N-1`, but `commit_min=N`.
```
```suggestion
            // In this scenario, our own DVC's commit is `N-1`, but `commit_min=N`.
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            // In this scenario, the our own `do_view_change`'s commit is `N-1`, but `commit_min=N`.
+            // Don't let the commit backtrack.
+            if (k.? < self.commit_min) {
+                assert(k.? == self.do_view_change_from_all_replicas[self.replica].?.header.commit);
```
Great!

sentientwaffle replied:
Ah, actually there was a bug in this assertion! It needs to handle committing multiple ops between while waiting for the DVC quorum.

sentientwaffle replied:
i.e. if we are replica 0, then replica 1's DVC commit_min might have been the highest in the quorum (`k`). But in the mean time we committed ahead of it (`self.commit_min > k`). I updated the comment, and the assert now reads:
```zig
                assert(self.commit_min >
                    self.do_view_change_from_all_replicas[self.replica].?.header.commit);
```

jorangreef replied:
Good catch. Expresses the concurrent `commit_min` increment more clearly now too.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2951,108 +2936,111 @@ pub fn ReplicaType(
+            assert(self.journal.header_with_op(self.op) != null);
+            assert(self.view_normal <= view_normal_canonical);
+
+            if (self.view_normal == view_normal_canonical) return self.op + 1;
```
Since when using this at the call site, we always subtract 1 (or add 1 to compare, i.e. in an assert), can we not rather return the op without the plus one? Or otherwise use an optional return type to indicate?

sentientwaffle replied:
Nice! Yes, it simplifies the caller to flip `op_uncanonical_min()` and return `op_canonical_max()` instead.

jorangreef replied:
Yes, `op_canonical_max()` would be perfect, then it's stated in the positive and we can drop the `uncanonical` naming.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2951,108 +2936,111 @@ pub fn ReplicaType(
+        /// 3. 8b is discarded due to the gap in 7.
+        /// 4. To distinguish between 6a and 6b (and safely discard 6a), the new leader trusts ops
+        ///    from the DVC(s) with the greatest `view_normal`.
+        fn op_uncanonical_min(self: *const Self, view_normal_canonical: u64) usize {
```
I was wondering whether we can't find a better name for this method?

Since the version of the op we have could in fact equally turn out to be canonical, we're just not sure.

sentientwaffle replied:
:thinking: My working definition for uncanonical headers is:

> An uncanonical message _may have_ been removed/changed by a prior view.

i.e. the new leader may decide to keep some "uncanonical" headers. By that definition I think the naming is correct. Are you using "canonical" to mean "all headers part of the new view, according to the leader at view start"?

jorangreef replied:
> Are you using "canonical" to mean "all headers part of the new view, according to the leader at view start"?

Yes. And these we know for sure to be canonical.

> My working definition for uncanonical headers is:

>> An uncanonical message may have been removed/changed by a prior view.

Yes. I'm thinking that it could also later be found to be canonical  say on another replica. So it's more a question of uncertainty. Not knowing for sure if it is canonical. But we can't say for sure that it is uncanonical.

jorangreef replied:
i.e. Like the opposite of a bloom filter, it will tell you for sure if something is not in a set, but it can't tell you for sure if something is in a set. The opposite of that. 😁

sentientwaffle replied:
Gotcha, we were using different definitions for "canonical headers" then. I was thinking of it in terms of the new leader's viewpoint, with the uncertainty built in. Phrased in the positive, my definition could have been "A canonical header was part of the latest view_normal".

(Even the "canonical" headers (according to this definition) may still be discarded if there is a gap across all DVCs.)

sentientwaffle replied:
What if we define "the highest view_normal before now" to be the "base view" (or similar — "base" is probably too general a term). Then `op_canonical_max()` can be named (e.g.) `op_base_max()`.

jorangreef replied:
> Phrased in the positive, my definition could have been "A canonical header was part of the latest view_normal".

I was actually thinking this too, and also that these are the same, since the headers from the latest `view_normal` will be installed by the new primary and be part of the `start_view`.

> Even the "canonical" headers (according to this definition) may still be discarded if there is a gap across all DVCs.

Yes, agreed.

sentientwaffle replied:
> I was actually thinking this too, and also that these are the same, since the headers from the latest view_normal will be installed by the new primary and be part of the start_view.

They will be installed, but will not necessarily be part of the `start_view`. A subset of the canonical headers (meaning here: headers from the latest `view_normal`) may be removed before the `start_view` (by `discard_uncommitted_headers`) if there is a gap in them (that the DVC quorum could not fill/repair). So I don't think they _are_ the same thing.

(I think this might be the root of our confusion here, so lmk if I should elaborate).

jorangreef replied:
Ah, I was not being precise enough. I was thinking they are the same _ceteris parabus_, but of course if we discard because of gaps then they will differ.

But _canonical_ is all w.r.t. to other possible candidate headers for the slots coming in from older views. That's how I was thinking of _canonical_. All about choosing which headers trump others, not so much about what will end up in the `start_view` message.

sentientwaffle replied:
Gotcha; that makes sense! I will expand on this in the code — it is a useful term to have when explaining/describing view change precedence.

Then for renaming `op_canonical_max()`, something like `leader_op_canonical_max()` (since it is only called by the new leader)?

jorangreef replied:
i.e. canonical headers may be uncommitted or committed, so they may still be discarded. They're canonical in the sense that they're ratified by being part of the DVC messages from the latest `view_normal`, and they help to decide which other headers are allowed in via hash chain connection.

Happy to think of another name also. I don't know if saying "base headers" is as clear, although "base" does mean a starting point. We need to capture the sense of "authoritative" or "approved" or "quorum". Maybe "quorum headers"?

jorangreef replied:
Let's keep it short with `op_canonical_max()` — we also have other methods only used by the primary, but it's fine if we don't prefix them.

It would be great to rename all the `leader` references to `primary` one day.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            //
+            // Removing these is necessary for correctness and liveness, to ensure that
+            // disconnected headers do not remain in place in lieu of gaps.
+            var op: u64 = op_uncanonical - 1;
```
Is it possible where a replica is so far behind that its uncanonical op is actually a few tens of ops behind, also completely disconnected, then we remove everything to the right, and then participate in a subsequent view change, but having lost what were committed headers (what we removed)?

On second thought, I guess not, since it's bounded by `pipeline_max` and we've noted this also in comments. But I thought I'd ask just to double check.

sentientwaffle replied:
Right, it is bounded by `pipeline_max`. Also if the header is possibly committed, it can't be disconnected at this point. The other DVCs would have provided the headers to repair the committed header's hash chain, due to the quorum intersection property.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            // Removing these is necessary for correctness and liveness, to ensure that
+            // disconnected headers do not remain in place in lieu of gaps.
+            var op: u64 = op_uncanonical - 1;
+            while (op <= self.op) : (op += 1) {
```
Can we refactor out some blocks within this function to keep the overall function body under 70 lines?

sentientwaffle replied:
I will refactor this some, but I don't think it will be under 70 lines yet. I have some other changes in mind for a future PR (to more aggressively verify the "all same-view_normal-DVC headers must match" invariant) that will take it the rest of the way. FWIW, I don't think this function actually does "too much" — rather, it has lots of documentation, checks, and logging, which inflates the line count. (Plus we have iterate the DVCs twice, both times filtering+checking `view_normal`).

One (possibly controversial :wink:) refactor — I restructured `discard_uncommitted_ops`:
- Renamed to `do_view_change_op_max`
- It now just returns the new `replica.op`; it does not set `replica.op` or truncate entries from the log.
- I moved the "search for hash break in uncanonical headers" into there as well.

My goal is to move as much of the state mutation as possible directly into `set_log_from_do_view_change_messages`, rather than the replica functions it invokes. (I find having the mutation as close together as possible makes it easier to mentally/manually check invariants).

jorangreef replied:
> My goal is to move as much of the state mutation as possible directly into set_log_from_do_view_change_messages, rather than the replica functions it invokes.

Excellent idea. Let's do it.

sentientwaffle committed:
VSR: Clarify canonical ops

- Replace `op_uncanonical_min()` with `op_canonical_max()`. (This is a simpler implementation, and closer to how the caller uses it).
- Replace `discard_uncommitted_headers()` with `do_view_change_op_max()`, to move mutation up a level (into `set_log_from_do_view_change_messages()`).
- Set the timestamp of the new leader before discarding any headers.
- Fix: In the `if (k.? < self.commit_min)` case, we can actually jump multiple commits, which means we may be jumping ahead of a DVC other than our own.

jorangreef commented on src/config.zig:
```diff
@@ -94,12 +94,20 @@ pub const connections_max = replicas_max + clients_max;
+const message_size_max_min = std.math.max(
+    sector_size,
+    std.mem.alignForward(
+        128 + pipeline_max * 128, // 128 == @sizeOf(vsr.Header)
```
Let's always use `@sizeOf(vsr.Header)` instead of a magic number, to always stay in sync through the type system with any Header size changes.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2972,75 +2983,26 @@ pub fn ReplicaType(
 
             // * When uncanonical_op_count = self.op - self.commit_min,
             //   self.op - (uncanonical_op_count - 1) =
             //   (self.op - uncanonical_op_count) + 1 =
```
Do we need to update any of these plus-ones or minus-ones?

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5263,9 +5218,92 @@ pub fn ReplicaType(
             };
         }
 
+        /// Identify uncommitted headers that to discard during a view change before the primary
```
"that to discard" -> "to discard"

Would be great also to edit to fit on one line, as a one line summary.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5263,9 +5218,92 @@ pub fn ReplicaType(
+        ///
+        /// Returns the highest op that:
+        /// - precedes any hash chain breaks in the uncanonical headers, and
+        /// - precedes any header gaps.
```
Nice summary.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5263,9 +5218,92 @@ pub fn ReplicaType(
+            assert(self.do_view_change_quorum);
+            assert(!self.repair_timeout.ticking);
+            assert(self.op >= self.commit_max);
+            assert(self.op - self.commit_max <= config.journal_slot_count);
```
Is there an off-by-one here? Since we are comparing an _index_ on the left with a _count_ on the right?

sentientwaffle replied:
:thinking: Borrowing the diagram from `op_checkpoint_next` (where journal_slot_count=8 and lsm_batch_multiple=2):

```
        ///     0   4   8   2   6   0   4  │ 0---4---
        ///   0 ─────✓·%                   │ 01234✓6%   initial log fill
        ///   1 ───────────✓·%             │ 890✓2%45   first wrap of log
```

When `op_checkpoint == 5`, the highest `replica.op` we can reach before another checkpoint (or even commit) is `13`. `13 - 5 = 8 = journal_slot_count`.

sentientwaffle replied:
This is an index vs a count, but we don't actually require that `commit_max` is in the log, i.e. counted by the slot_count (when `commit_max` is equal to `op_checkpoint`). Or rather, we _won't_ require that — I'm fixing that issue in a different branch (via `VSRState.commit_min_checksum`) :sweat_smile:.

jorangreef replied:
Great!

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5263,9 +5218,92 @@ pub fn ReplicaType(
+        /// the rest of the cluster may have already discarded it. We therefore iterate over our
+        /// uncommitted header gaps to discard any that may be impossible to repair.
+        ///
+        /// For example, if the old primary replicates ops=7,8,9 (all uncommitted) but only op=9 is
+        /// prepared on another replica before the old primary crashes, then this function finds a
+        /// gap for ops=7,8 and will attempt to discard ops 7,8,9.
+        fn do_view_change_op_max(self: *const Self, op_canonical: u64) u64 {
```
Great comment.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5263,9 +5218,92 @@ pub fn ReplicaType(
+                var op: u64 = op_canonical;
+                while (op <= self.op) : (op += 1) {
+                    if (self.journal.header_with_op(op)) |header| {
+                        if (self.journal.header_with_op(op + 1)) |next| {
```
I think we might be going too far here in the loop, since on the last iteration `op=self.op`  and then we look for the next op, i.e. `self.op+1`?

jorangreef replied:
i.e. I would have expected that with `self.op` set, we don't have any headers beyond `self.op` at this stage in the WAL?

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5263,9 +5218,92 @@ pub fn ReplicaType(
+
+                var op = op_committed + 1;
+                while (op <= self.op) : (op += 1) {
+                    if (self.journal.header_with_op(op) == null) break :blk op - 1;
```
It would be great to use the same style w.r.t. where we do the plus-one as we do for `op_before_break`.

i.e. `op_before_break` assigns `var op: u64 = op_canonical;` and then adds one when looking for the next header, wherease `op_before_gap` assigns `var op = op_committed + 1;` and then corrects for this in the loop.

I think we could make both loops symmetrical? And then the first style seems better.

sentientwaffle replied:
Good point! It is cleaner that way.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5263,9 +5218,92 @@ pub fn ReplicaType(
+                } else break :blk self.op;
+            };
+
+            return std.math.min(op_before_break, op_before_gap);
```
Nice way to work this whole function, getting to this point.

jorangreef reviewed (commented):
This new refactor is much cleaner!

jorangreef commented:
The `do_view_change_op_max()` strategy is great.

sentientwaffle committed:
VSR: Fix repair of root prepare during recovery

`replace_headers()` will never repair a (corrupt) root prepare (op=0) because `op=0 ≤ op_checkpoint=0`.
But `on_recovery_response` must still repair it before asserting the "replica.op exists" invariant, since `replica.op` might be 0.
It is also good to move that repair before the transition to normal status.

(Also includes misc cleanup.)

jorangreef reviewed (commented):
LGTM!

I can't Approve because I'm the PR author.
